### ISSUE TITLE
Memory leak in zipnn_core C extension causes OOM in batch processing

### ISSUE THREAD

**Issue by DaiShiResearch** [zipnn/zipnn#78]
# Memory leak in zipnn_core C extension causes OOM in batch processing

## Description

The `zipnn_core` C extension has critical memory leaks in both compression (`py_zipnn_core`) and decompression (`py_combine_dtype`) functions. Memory grows proportionally to the size of data being compressed/decompressed, and is never released.

**This is a severe issue for production workloads**: In large-scale data processing pipelines, memory accumulates continuously until the system runs out of RAM, resulting in OOM (Out of Memory).

## Reproduce

```python
import torch
from zipnn import ZipNN

zpn = ZipNN(method='huffman', input_format='torch')
for i in range(100):
    tensor = torch.randn(1000, 1000, dtype=torch.float16)  # ~1.9 MB per tensor
    compressed = zpn.compress(tensor)
    decompressed = zpn.decompress(compressed)
    del compressed, decompressed, tensor
    # Memory grows ~6MB per iteration (3x tensor size) instead of staying stable
```

## Actual Behavior

```
Total data processed: 190.7 MB
Memory increase: 608.2 MB (3.19x of data processed)

Initial:  505 MB → Final: 1113 MB (+608 MB for 100 iterations)
```

## Root Cause

In `csrc/zipnn_core.c`:
1. `resultBuf` (malloc'd) never freed after `PyMemoryView_FromBuffer`
2. `Py_buffer` references from `PyArg_ParseTuple` never released

## Fix

Fixes #77 

See https://github.com/zipnn/zipnn/pull/77

After fix: Memory leak is completely resolved. Memory remains stable after initial warmup, enabling reliable operation in large-scale batch processing workloads.

**Comment by moshik1** [comment]
Thanks a lot @DaiShiResearch for the detailed issue report and the PR. I’ll take a look later this week.

**Comment by DaiShiResearch** [comment]
@moshik1 Hi, just a gentle follow-up as it’s been a few months. Would anyone be available to review this PR?

**Comment by moshik1** [comment]
Hi @DaiShiResearch sorry for the delayed response, you were absolutely right to flag it. 
The memory leak you identified is real and the reproduction data is excellent.

We reviewed the fix in #77 and left a detailed comment there. The root cause analysis is spot on.
We have one suggestion before merging: preserving zero-copy via `PyCapsule` instead of `PyBytes`,
since both functions operate on large tensors and we want to keep zipnn's performance profile intact,
keeping the repo fast is something we care deeply about.

Overall direction is correct and we want to get this merged. 
Thanks for the contribution  it's genuinely appreciated!

**Comment by DaiShiResearch** [comment]
@moshik1 Thanks for confirming! I've replied on #77 with detailed findings on the PyCapsule approach — the capsule reference counting works correctly in isolation, but in practice the leak persists due to how torch.from_numpy() manages buffer references at the C++ level. Full test data is in the PR comment.

The current PyBytes/PyByteArray fix reliably resolves the OOM. Achieving true zero-copy may require a different approach (e.g. PyArray_SimpleNewFromData), which would involve larger changes across the C and Python layers — happy to explore that in a follow-up PR if you'd like.

**Comment by DaiShiResearch** [comment]
@moshik1 Quick update — the zero-copy follow-up is done and pushed to the PR branch.

The new approach completely eliminates the memcpy overhead while keeping the leak fix intact. Instead of the PyCapsule path (which leaked due to the `tensor → numpy → memoryview → capsule` reference chain), we pre-allocate the Python buffer object first (`PyByteArray_FromStringAndSize(NULL, size)`) and have the C worker threads write directly into it. No intermediate objects, no separate `malloc`, no copy. Python GC owns the memory from the start.

Results on 1GB BF16 (10 iterations):

| Metric | PyPI v0.5.3 (leaking) | This PR |
|---|---|---|
| Compress | 1246 MB/s | **1287 MB/s** |
| Decompress | 15405 MB/s | **15813 MB/s** |
| Memory leak | ❌ | ✅ Fixed |

All existing tests pass. Only `csrc/zipnn_core.c` was modified.

This should be ready for final review. Would be great to get it into the next PyPI release since the OOM issue is hitting production users — happy to help with anything needed to move it forward.

**Comment by moshik1** [comment]
@DaiShiResearch sounds amazing. I will review it on Sunday.

**Comment by moshik1** [comment]
Hi @DaiShiResearch,
Approved, great work! If you’d like to continue contributing, that would be great.
Thanks a lot!
I’ll publish a new PyPI release next week and then close the issue.

**Comment by DaiShiResearch** [comment]
Hi @moshik1,
Thanks a lot for the quick review and for planning the upcoming PyPI release — really appreciate you pushing this forward, especially with the OOM issue affecting production users.

Glad to hear the PR looks good on your side. I’d definitely be happy to continue contributing going forward and help improve the project wherever I can.

Looking forward to the release!

**Comment by moshik1** [comment]
Hi @DaiShiResearch,
I uploaded a PyPI version. Please take a look

Thanks! 
Moshik
